File: AnyDaf/StudyModels.swift
import Foundation

// MARK: - YCT Library Models

/// Which YCT site an article comes from
enum YCTSource: String, Codable {
    case library // library.yctorah.org
    case psak    // psak.yctorah.org
}

/// How closely an article matches the current daf
enum ResourceMatchType: Equatable, Codable {
    case exact(daf: Int)          // matches current daf exactly
    case nearby(daf: Int)         // matches a nearby daf (±2)
    case tractateWide(daf: Int)   // matches tractate; daf is the article's specific daf reference

    var referencedDaf: Int {
        switch self {
        case .exact(let d), .nearby(let d), .tractateWide(let d): return d
        }
    }

    // MARK: Codable — manual implementation required for enums with associated values

    private enum CodingKeys: String, CodingKey { case type, daf }
    private enum TypeName: String, Codable { case exact, nearby, tractateWide }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .exact(let d):        try c.encode(TypeName.exact,        forKey: .type); try c.encode(d, forKey: .daf)
        case .nearby(let d):       try c.encode(TypeName.nearby,       forKey: .type); try c.encode(d, forKey: .daf)
        case .tractateWide(let d): try c.encode(TypeName.tractateWide, forKey: .type); try c.encode(d, forKey: .daf)
        }
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let type_ = try c.decode(TypeName.self, forKey: .type)
        let daf   = try c.decode(Int.self,      forKey: .daf)
        switch type_ {
        case .exact:        self = .exact(daf: daf)
        case .nearby:       self = .nearby(daf: daf)
        case .tractateWide: self = .tractateWide(daf: daf)
        }
    }
}

/// A taxonomy term from the YCT Library reference hierarchy
struct YCTReferenceTerm: Identifiable {
    let id: Int
    let name: String
    let slug: String
    let parent: Int
}

/// An article/essay from the YCT Torah Library
struct YCTArticle: Identifiable, Codable {
    let id: Int
    let title: String
    let excerpt: String
    let date: String
    let link: String
    let authorName: String
    var matchType: ResourceMatchType
    /// Additional daf references beyond the primary one (sorted ascending).
    var additionalDafs: [Int]
    /// Which YCT site this article came from.
    var source: YCTSource

    init(id: Int, title: String, excerpt: String, date: String, link: String,
         authorName: String, matchType: ResourceMatchType, additionalDafs: [Int] = [],
         source: YCTSource = .library) {
        self.id = id; self.title = title; self.excerpt = excerpt; self.date = date
        self.link = link; self.authorName = authorName; self.matchType = matchType
        self.additionalDafs = additionalDafs; self.source = source
    }

    // Backward-compatible decoder: old cache files lack `additionalDafs` and `source`.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id             = try c.decode(Int.self,               forKey: .id)
        title          = try c.decode(String.self,            forKey: .title)
        excerpt        = try c.decode(String.self,            forKey: .excerpt)
        date           = try c.decode(String.self,            forKey: .date)
        link           = try c.decode(String.self,            forKey: .link)
        authorName     = try c.decode(String.self,            forKey: .authorName)
        matchType      = try c.decode(ResourceMatchType.self, forKey: .matchType)
        additionalDafs = try c.decodeIfPresent([Int].self,    forKey: .additionalDafs) ?? []
        source         = try c.decodeIfPresent(YCTSource.self, forKey: .source) ?? .library
    }
}

// MARK: - Study Models

/// Scope of study: one side or full daf
enum StudyScope: String, CaseIterable {
    case amudA = "Amud A"
    case amudB = "Amud B"
    case fullDaf = "Full Daf"
}

/// Quiz / question format — stored in AppStorage as a raw String
enum QuizMode: String, CaseIterable {
    case multipleChoice = "multipleChoice"
    case flashcard      = "flashcard"
    case fillInBlank    = "fillInBlank"
    case shortAnswer    = "shortAnswer"

    var displayName: String {
        switch self {
        case .multipleChoice: return "Multiple Choice"
        case .flashcard:      return "Flashcard"
        case .fillInBlank:    return "Fill in the Blank"
        case .shortAnswer:    return "Short Answer"
        }
    }

    var modeDescription: String {
        switch self {
        case .multipleChoice:
            return "Choose the correct answer from four options."
        case .flashcard:
            return "Read the question, think of your answer, then reveal and self-grade."
        case .fillInBlank:
            return "Type your answer. Claude grades it, accepting also Aramaic and Hebrew input."
        case .shortAnswer:
            return "Answer in your own words. Claude grades it, accepting paraphrases and alternate formulations."
        }
    }
}

/// How the original Hebrew/Aramaic source text is displayed alongside the English translation.
/// Stored in AppStorage as a raw String.
enum SourceDisplayMode: String, CaseIterable {
    case toggle  = "toggle"
    case stacked = "stacked"

    var displayName: String {
        switch self {
        case .toggle:  return "Toggle"
        case .stacked: return "Top & Bottom"
        }
    }

    var modeDescription: String {
        switch self {
        case .toggle:  return "Tap a button to switch between source text and translation."
        case .stacked: return "Each paragraph shown as source above translation, scroll through paired paragraphs."
        }
    }
}

/// Result from AI-graded fill-in-the-blank answer
struct GradeResult: Equatable {
    let isCorrect: Bool
    let feedback: String
}

/// A single quiz question — supports multiple choice, flashcard, and fill-in-blank
struct QuizQuestion: Identifiable {
    let id = UUID()
    let mode: QuizMode
    let question: String
    let correctAnswer: String           // used by all modes

    // ── Multiple Choice ────────────────────────────────────────
    let choices: [String]               // exactly 4 for MC; empty otherwise
    let correctIndex: Int               // 0–3 for MC; -1 for other modes
    var selectedIndex: Int?             // nil until answered

    // ── Flashcard ──────────────────────────────────────────────
    var selfMarkedCorrect: Bool?        // nil until self-graded

    // ── Fill-in-blank ──────────────────────────────────────────
    var userText: String?               // what the user typed
    var isGrading: Bool = false
    var gradeResult: GradeResult?

    // MARK: - Computed

    var isAnswered: Bool {
        switch mode {
        case .multipleChoice:            return selectedIndex != nil
        case .flashcard:                 return selfMarkedCorrect != nil
        case .fillInBlank, .shortAnswer: return gradeResult != nil
        }
    }

    var isCorrect: Bool {
        switch mode {
        case .multipleChoice:            return selectedIndex == correctIndex
        case .flashcard:                 return selfMarkedCorrect == true
        case .fillInBlank, .shortAnswer: return gradeResult?.isCorrect == true
        }
    }

    // MARK: - Factory methods

    static func multipleChoice(question: String,
                               choices: [String],
                               correctIndex: Int) -> QuizQuestion {
        QuizQuestion(mode: .multipleChoice,
                     question: question,
                     correctAnswer: choices[correctIndex],
                     choices: choices,
                     correctIndex: correctIndex)
    }

    static func flashcard(question: String, answer: String) -> QuizQuestion {
        QuizQuestion(mode: .flashcard,
                     question: question,
                     correctAnswer: answer,
                     choices: [],
                     correctIndex: -1)
    }

    static func fillInBlank(question: String, answer: String) -> QuizQuestion {
        QuizQuestion(mode: .fillInBlank,
                     question: question,
                     correctAnswer: answer,
                     choices: [],
                     correctIndex: -1)
    }

    static func shortAnswer(question: String, answer: String) -> QuizQuestion {
        QuizQuestion(mode: .shortAnswer,
                     question: question,
                     correctAnswer: answer,
                     choices: [],
                     correctIndex: -1)
    }
}

/// One logical section of the daf (e.g., Mishna, Gemara, or an untitled block)
struct StudySection: Identifiable {
    let id = UUID()
    let title: String            // "MISHNA", "GEMARA", etc.
    let rawText: String          // joined English HTML (used by quiz, summary, read-aloud)
    let rawSegments: [String]    // individual English HTML segments (for paragraph-aligned display)
    let hebrewText: String?      // joined Hebrew/Aramaic HTML (used by toggle mode)
    let hebrewSegments: [String] // individual Hebrew HTML segments (for paragraph-aligned display)
    var summary: String?         // Claude-generated summary
    var quizQuestions: [QuizQuestion]
}

/// Full study session state
struct StudySession {
    let tractate: String
    let daf: Int
    let scope: StudyScope
    var sections: [StudySection]
    var currentSectionIndex: Int = 0
    var amudBSectionIndex: Int?
    /// Tail of the previous daf, shown in brackets when amud a starts mid-sentence.
    var precedingContext: String? = nil
    /// Head of the next daf, shown in brackets when amud b ends mid-sentence.
    var followingContext: String? = nil

    var currentSection: StudySection? {
        guard currentSectionIndex < sections.count else { return nil }
        return sections[currentSectionIndex]
    }

    var isComplete: Bool {
        currentSectionIndex >= sections.count
    }

    var progress: Double {
        guard !sections.isEmpty else { return 0 }
        return Double(currentSectionIndex) / Double(sections.count)
    }
}

File: android/app/src/main/java/com/anydaf/model/StudyModels.kt
package com.anydaf.model

import java.util.UUID

// YCT Library Models

enum class YCTSource { LIBRARY, PSAK }

sealed class ResourceMatchType {
    data class Exact(val daf: Int) : ResourceMatchType()
    data class Nearby(val daf: Int) : ResourceMatchType()
    data class TractateWide(val daf: Int) : ResourceMatchType()

    val referencedDaf: Int get() = when (this) {
        is Exact -> daf
        is Nearby -> daf
        is TractateWide -> daf
    }
}

data class YCTReferenceTerm(
    val id: Int,
    val name: String,
    val slug: String,
    val parent: Int
)

data class YCTArticle(
    val id: Int,
    val title: String,
    val excerpt: String,
    val date: String,
    val link: String,
    val authorName: String,
    val matchType: ResourceMatchType = ResourceMatchType.Exact(0),
    /** Additional daf references beyond the primary one (sorted ascending). */
    val additionalDafs: List<Int> = emptyList(),
    /** Which YCT site this article came from. */
    val source: YCTSource = YCTSource.LIBRARY
)

// Study Models

enum class StudyScope(val displayName: String) {
    AMUD_A("Amud A"),
    AMUD_B("Amud B"),
    FULL_DAF("Full Daf")
}

enum class StudyMode { FACTS, CONCEPTUAL }

enum class QuizMode(val displayName: String, val description: String) {
    MULTIPLE_CHOICE(
        "Multiple Choice",
        "Choose the correct answer from four options."
    ),
    FLASHCARD(
        "Flashcard",
        "Read the question, think of your answer, then reveal and self-grade."
    ),
    FILL_IN_BLANK(
        "Fill in the Blank",
        "Type your answer. Claude grades it, accepting also Aramaic and Hebrew input."
    ),
    SHORT_ANSWER(
        "Short Answer",
        "Answer in your own words. Claude grades it, accepting paraphrases and alternate formulations."
    )
}

enum class SourceDisplayMode(val displayName: String, val description: String) {
    TOGGLE("Toggle", "Tap a button to switch between source text and translation."),
    STACKED("Top & Bottom", "Each paragraph shown as source above translation, scroll through paired paragraphs.")
}

enum class StudyFontSize(val displayName: String, val spSize: Float, val articleFontPx: Int) {
    X_SMALL("Extra Small", 12f, 14),
    SMALL("Small", 15f, 17),
    MEDIUM("Medium", 18f, 20),
    LARGE("Large", 21f, 23),
    X_LARGE("Extra Large", 26f, 28);

    companion object {
        /** X_SMALL is tablet-only — it maps to iOS system default size, already small on phones. */
        fun displayEntries(isTablet: Boolean): List<StudyFontSize> =
            if (isTablet) entries else entries.filter { it != X_SMALL }
    }
}

data class GradeResult(
    val isCorrect: Boolean,
    val feedback: String
)

data class QuizQuestion(
    val id: UUID = UUID.randomUUID(),
    val mode: QuizMode,
    val question: String,
    val correctAnswer: String,
    // Multiple choice
    val choices: List<String> = emptyList(),
    val correctIndex: Int = -1,
    var selectedIndex: Int? = null,
    // Flashcard
    var selfMarkedCorrect: Boolean? = null,
    // Fill-in-blank / Short answer
    var userText: String? = null,
    var isGrading: Boolean = false,
    var gradeResult: GradeResult? = null
) {
    val isAnswered: Boolean get() = when (mode) {
        QuizMode.MULTIPLE_CHOICE -> selectedIndex != null
        QuizMode.FLASHCARD -> selfMarkedCorrect != null
        QuizMode.FILL_IN_BLANK, QuizMode.SHORT_ANSWER -> gradeResult != null
    }

    val isCorrect: Boolean get() = when (mode) {
        QuizMode.MULTIPLE_CHOICE -> selectedIndex == correctIndex
        QuizMode.FLASHCARD -> selfMarkedCorrect == true
        QuizMode.FILL_IN_BLANK, QuizMode.SHORT_ANSWER -> gradeResult?.isCorrect == true
    }

    companion object {
        fun multipleChoice(question: String, choices: List<String>, correctIndex: Int) =
            QuizQuestion(
                mode = QuizMode.MULTIPLE_CHOICE,
                question = question,
                correctAnswer = choices.getOrElse(correctIndex) { "" },
                choices = choices,
                correctIndex = correctIndex
            )

        fun flashcard(question: String, answer: String) =
            QuizQuestion(mode = QuizMode.FLASHCARD, question = question, correctAnswer = answer)

        fun fillInBlank(question: String, answer: String) =
            QuizQuestion(mode = QuizMode.FILL_IN_BLANK, question = question, correctAnswer = answer)

        fun shortAnswer(question: String, answer: String) =
            QuizQuestion(mode = QuizMode.SHORT_ANSWER, question = question, correctAnswer = answer)
    }
}

data class StudySection(
    val id: UUID = UUID.randomUUID(),
    val title: String,
    val rawText: String,
    val rawSegments: List<String>,
    val hebrewText: String?,
    val hebrewSegments: List<String>,
    var summary: String? = null,
    var quizQuestions: List<QuizQuestion> = emptyList()
)

data class StudySession(
    val tractate: String,
    val daf: Int,
    val scope: StudyScope,
    val sections: List<StudySection>,
    val currentSectionIndex: Int = 0,
    val amudBSectionIndex: Int? = null,
    val precedingContext: String? = null,
    val followingContext: String? = null
) {
    val currentSection: StudySection? get() =
        sections.getOrNull(currentSectionIndex)

    val isComplete: Boolean get() =
        currentSectionIndex >= sections.size

    val progress: Double get() =
        if (sections.isEmpty()) 0.0
        else currentSectionIndex.toDouble() / sections.size
}
